resource/ui/ts/models/timeseries.ts

```diff
@@ -43,68 +44,183 @@ module Models {
-                            r.datapoints = []
-                        }
-                    });
-                    return d;
-                });
+                }
+            }
+
+            /**
+             * Avg instantiates a new AvgSelector for the supplied time series.
+             */
+            export function Avg(series:string):AvgSelector {
+                return new AvgSelector(series);
+            }
+
+            /**
+             * AvgRate instantiates a new AvgRateSelector for the supplied time
+             * series.
+             */
+            export function AvgRate(series:string):AvgRateSelector {
+                return new AvgRateSelector(series);
+            }
         }
 
-        /** 
-         * RecentQuery is a query which monitors a duration of constant size
-         * extending backwards from the current time. When refreshed, the query
-         * will be re-issued to extend to the current time.
+        /**
+         * time contains available time span specifiers for metrics queries.
          */
-        export class RecentQuery {
-            private _series:string[];
-            constructor(public windowDuration:number, private _agg:QueryAggregator, ...series:string[]) {
-                this._series = series;
+        export module time {
+            /**
+             * TimeSpan is the interface implemeted by time span specifiers.
+             */
+            export interface TimeSpan {
+                /**
+                 * timespan returns a two-value number array which defines the
+                 * time range of a query. The first value is a timestamp for the
+                 * start of the range, the second value a timestamp for the end
+                 * of the range.
+                 */
+                timespan():number[];
             }
 
-            query():promise<QueryResultSet> {
-                var endTime = new Date();
-                var startTime = new Date(endTime.getTime() - this.windowDuration);
-                return query(startTime, endTime, this._agg, this._series);
+            export function Recent(duration:number):TimeSpan {
+                return {
+                    timespan: function():number[] {
+                        var endTime = new Date();
+                        var startTime = new Date(endTime.getTime() - duration);
+                        return [startTime.getTime(), endTime.getTime()]
+                    }
+                }
             }
         }
 
         /**
+         * Query is a class which describes a single query to the time series
+         * server.
+         */
+        class Query {
+            constructor(private _selectors:select.Selector[]) {}
+
+            timespan = Utils.chainProp(this, time.Recent(10 * 60 * 1000));
+
+            execute():promise<QueryResultSet> {
+                var s = this.timespan().timespan();
+                var req:QueryRequestSet = {
+                    start_nanos: s[0] * 1.0e6,
```

Comment:
I wonder if we should have a micro to nano util function somewhere.  We do this a lot.